### PR TITLE
[WB-5255][CLI-864][WB-5437] Console redirect perf + bug fixes

### PR DESCRIPTION
**Pull request by farizrahman4u**
Fixes: https://github.com/ultralytics/yolov5/issues/2685
Fixes: #2097 

https://wandb.atlassian.net/browse/WB-5255
https://wandb.atlassian.net/browse/CLI-864
https://wandb.atlassian.net/browse/WB-5437

Description
-----------

What does the PR do?

- Fuse char level loops using regex and numpy
- Optimize color & style decoding from emulator state
- Dispatch emulator write to a different thread, unblocking sys.stdout/err.write
- Reduce output volume for offline runs in Windows and notebooks
- Fix output truncating issue when script is run as a subproc in jupyter notebook (`!python train.py`)
- Some clean up + minor bug fixes

Testing
-------

Unit tests

### CODE REVIEW

File: wandb/sdk/lib/redirect.py [hunk 1]
```diff
@@ -27,6 +29,19 @@
 ANSI_CSI_RE = re.compile("\001?\033\\[((?:\\d|;)*)([a-zA-Z])\002?")
 ANSI_OSC_RE = re.compile("\001?\033\\]([^\a]*)(\a)\002?")
 
+SEP_RE = re.compile(
```
Comment (vanpelt) [inline wandb/sdk/lib/redirect.py:32]: Can you add a comment explaining what this is / where you found it?

File: wandb/sdk/lib/redirect.py [hunk 2]
```diff
@@ -621,7 +658,7 @@ def uninstall(self):
             return
         self._installed = False
         self.src_wrapped_stream.flush()
-        time.sleep(1)
+        time.sleep(0.5)
```
Comment (vanpelt) [inline wandb/sdk/lib/redirect.py:661]: Why are we sleeping at all?  If there's a deterministic reason let's explain it here.  If there are cases where the sleep won't work we should find a solution that always works.

File: wandb/sdk/lib/ipython.py [hunk 3]
```diff
@@ -13,13 +14,29 @@
...
 
 
+def _is_jupyter_process():
+    """
+    Detect if a script is run as a sub process (!python train.py) in a jupyter environment.
+    The script could start its own sub procs and this method would return True for all of them.
+    """
+    p = psutil.Process()
+    while p:
+        if p.name().startswith("jupyter-note"):
```
Comment (vanpelt) [inline wandb/sdk/lib/ipython.py:24]: What makes us sure all ipython sessions will have this name?

File: requirements.txt [hunk 4]
```diff
@@ -15,3 +15,4 @@ PyYAML
 typing; python_version < '3.5'
 enum34; python_version < '3.4'
 pathtools # supports vendored version of watchdog 0.9.0
+numpy
```
Comment (vanpelt) [inline requirements.txt:18]: Why are we making numpy a hard dependency?

Comment (farizrahman4u) [reply requirements.txt:18]: Used in redirect.py.

File: wandb/sdk/lib/redirect.py (hunk 1 above)
Comment (farizrahman4u) [reply wandb/sdk/lib/redirect.py:33]: Simplified and comment added.

File: wandb/sdk/lib/ipython.py (hunk 3 above)
Comment (farizrahman4u) [reply wandb/sdk/lib/ipython.py]: Looking at better solutions...

Comment (farizrahman4u) [reply wandb/sdk/lib/ipython.py]: Done.

File: requirements.txt (hunk 4 above)
Comment (farizrahman4u) [reply requirements.txt]: Hard dependency removed. Falls back to naïve impl if numpy not available.

File: wandb/sdk/lib/redirect.py [hunk 5]
```diff
@@ -620,13 +688,19 @@ def uninstall(self):
...
-        self._stopped.set()
+        time.sleep(0.1)
         os.dup2(self._orig_src_fd, self.src_fd)
         os.write(self._pipe_write_fd, b"\n")
         os.close(self._pipe_write_fd)
         os.close(self._pipe_read_fd)
+
+        # Joining daemonic thread might hang, so we wait for the queue to empty out instead:
+        while not self._queue.empty():
+            time.sleep(0.1)
```
Comment (vanpelt) [inline wandb/sdk/lib/redirect.py:701]: Is there a scenario where the queue never emptys?

File: wandb/sdk/lib/redirect.py [hunk 6]
```diff
@@ -620,13 +688,19 @@ def uninstall(self):
         if not self._installed:
             return
         self._installed = False
+
         self.src_wrapped_stream.flush()
-        time.sleep(1)
-        self._stopped.set()
+        time.sleep(0.1)
```
Comment (vanpelt) [inline wandb/sdk/lib/redirect.py:693]: If we really need to add sleeps let's add a comment describing why it's required and why we chose 0.1

File: wandb/sdk/lib/redirect.py [hunk 7]
```diff
@@ -347,28 +379,40 @@ def _handle_csi(self, csi, params, command):
...
-                for k in c.__slots__[3:]:
-                    ck = c[k]
-                    if ck != prev_char[k]:
-                        if not ck:
-                            k = "/" + k
-                        out += _get_char(ANSI_STYLES_REV[k])
-                out += c.data
-                prev_char = c
-            return out
+            out = [line[i].data for i in range(line_len)]
```
Comment (vanpelt) [inline wandb/sdk/lib/redirect.py:386]: The code below is GNARLEY.  Let's add some more comments about what is happening here.

File: wandb/sdk/lib/redirect.py (hunk 5 above)
Comment (farizrahman4u) [reply wandb/sdk/lib/redirect.py:701]: Hmm.. in theory no. But lets bail after a while just in case?

Comment (vanpelt) [reply wandb/sdk/lib/redirect.py:701]: Yeah, I suppose we could do this for 10 seconds and give up with a warning if we ever take that long.  Dropping log output isn't the end of the world.

**Review by vanpelt — APPROVED** [review]
Nice!  Love the comment.

**Review by vanpelt — COMMENTED** [review]
I'm curious what actually fixed the memory leak?

File: wandb/sdk/lib/redirect.py [hunk 8]
```diff
@@ -620,22 +706,35 @@ def uninstall(self):
...
+            target=self.src_wrapped_stream.flush
+        )  # Calling flush() from the current thread does not flush the buffer instantly.
+        t.start()
+        t.join(timeout=10)
+
+        self._emulator_write_thread.join(timeout=5)
+        if self._emulator_write_thread.is_alive():
+            wandb.termlog("Processing terminal ouput (%s)..." % self.src)
+            self._emulator_write_thread.join()
+            wandb.termlog("Done.")
```
[the remark this answers is not in the document]
Comment (farizrahman4u) [reply wandb/sdk/lib/redirect.py:752]: We are printing it only if it takes more than 5 seconds. If there is too much data to process, emulator write thread would bail on its own and upload the raw stream chars (without running it through the emulator)..

File: wandb/sdk/lib/redirect.py [hunk 9]
```diff
@@ -397,6 +469,15 @@ def read(self):
                     )
                     + os.linesep
                 )
+        if num_lines > self._MAX_LINES:
```
Comment (farizrahman4u) [inline wandb/sdk/lib/redirect.py:472]: @vanpelt This is the mem leak fix

**Pull request merged**